app/v1/services/EnrolmentsAuthService.scala

```diff
@@ -38,18 +40,55 @@ class EnrolmentsAuthService @Inject()(val connector: AuthConnector) {
     override def authConnector: AuthConnector = connector
   }
 
-  def authorised(predicate: Predicate)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[AuthOutcome] = {
-    authFunction.authorised(predicate).retrieve(affinityGroup and allEnrolments) {
-      case Some(Individual) ~ enrolments => createUserDetailsWithLogging(affinityGroup = "Individual", enrolments)
-      case Some(Organisation) ~ enrolments => createUserDetailsWithLogging(affinityGroup = "Organisation", enrolments)
-      case Some(Agent) ~ enrolments => createUserDetailsWithLogging(affinityGroup = "Agent", enrolments)
-      case _ =>
-        Logger.warn(s"[AuthorisationService] [authoriseAsClient] Authorisation failed due to unsupported affinity group.")
-        Future.successful(Left(LegacyUnauthorisedError))
-    } recoverWith unauthorisedError
+  def authorised(predicate: Predicate, nrsRequired: Boolean = false)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[AuthOutcome] = {
+    if(!nrsRequired){
+      authFunction.authorised(predicate).retrieve(affinityGroup and allEnrolments) {
+        case Some(Individual) ~ enrolments => createUserDetailsWithLogging(affinityGroup = "Individual", enrolments)
+        case Some(Organisation) ~ enrolments => createUserDetailsWithLogging(affinityGroup = "Organisation", enrolments)
+        case Some(Agent) ~ enrolments => createUserDetailsWithLogging(affinityGroup = "Agent", enrolments)
+        case _ =>
+          Logger.warn(s"[AuthorisationService] [authoriseAsClient] Authorisation failed due to unsupported affinity group.")
+          Future.successful(Left(LegacyUnauthorisedError))
+      }recoverWith unauthorisedError
+    } else {
+      authFunction.authorised(predicate).retrieve(affinityGroup and allEnrolments
+        and internalId and externalId and agentCode and credentials
+        and confidenceLevel and nino and saUtr and name and dateOfBirth
+        and email and agentInformation and groupIdentifier and credentialRole
+        and mdtpInformation and credentialStrength and loginTimes
+      ) {
+        case affGroup ~ enrolments ~ inId ~ exId ~ agCode ~ creds
+          ~ confLevel ~ ni ~ saRef ~ nme ~ dob
+          ~ eml ~ agInfo ~ groupId ~ credRole
+          ~ mdtpInfo ~ credStrength ~ logins
+          if affGroup.contains(AffinityGroup.Organisation) || affGroup.contains(AffinityGroup.Individual) || affGroup.contains(AffinityGroup.Agent) =>
```

Comment:
could this be replaced by something like supportedAffGroup.contains(affGroup) where supportedAffGroup is Seq(AffinityGroup.Organisation, AffinityGroup.Individual, AffinityGroup.Agent)

Reply:
can do but here **affGroup** is an Option. So we need to call .getOrElse. or I can do match case for affGroup like
 case Some(Individual)  => 
 case Some(Organisation) => 
 case Some(Agent) =>